### ISSUE TITLE
aws_kinesisanalyticsv2 - destroy still not removing ENI's consistently

### ISSUE THREAD

**Issue by pmckl** [hashicorp/terraform-provider-aws#25528]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform CLI and Terraform AWS Provider Version

terraform version: 0.15.5
hashicorp/aws: v3.75.2

### Affected Resource(s)

* aws_kinesisanalyticsv2

### Terraform Configuration Files

```hcl
resource "aws_security_group" "security_group" {
  name        = local.security_group_name
  description = "KDA Security group for ${var.project}"
  vpc_id      = var.vpc_id == "" ? data.aws_vpc.vpc[0].id : var.vpc_id
  egress {
    from_port        = 0
    to_port          = 0
    protocol         = "-1"
    cidr_blocks      = ["0.0.0.0/0"]
    ipv6_cidr_blocks = ["::/0"]
  }
  tags = local.common_tags
}
```
```hcl
resource "aws_kinesisanalyticsv2_application" "kda" {
  name                   = local.application_name
  runtime_environment    = var.runtime_environment
  service_execution_role = module.kda_role.iam_role_arn
  start_application      = false
  depends_on = [
    aws_s3_bucket_object.placeholder
  ]
  lifecycle {
    ignore_changes = [
      application_configuration.0.application_code_configuration,
      application_configuration.0.environment_properties,
    ]
  }
  cloudwatch_logging_options {
    log_stream_arn = aws_cloudwatch_log_stream.this.arn
  }
  application_configuration {
    run_configuration {
      application_restore_configuration {
        application_restore_type = var.application_restore_type
        snapshot_name            = var.snapshot_name
      }
      flink_run_configuration {
        allow_non_restored_state = var.allow_non_restored_state
      }
    }
    application_snapshot_configuration {
      snapshots_enabled = var.snapshots_enabled
    }
...
    vpc_configuration {
      security_group_ids = [aws_security_group.security_group.id]
      subnet_ids         = length(var.subnet_ids) == 0 ? data.aws_subnet_ids.subnet_id[0].ids : var.subnet_ids
    }
    ...
    }
  }
}
```
### Debug Output

None

### Panic Output

None

### Expected Behavior

When destroying aws_kinesisanalyticsv2, it should destroy the ENIs as well.

### Actual Behavior

1 out of 10 times, the ENIs are not destroyed, which prevents the aws_security_group destruction.

### Steps to Reproduce

terraform apply
start the KDA application
terraform destroy

### Important Factoids

None

### References

*[ #7600](https://github.com/hashicorp/terraform-provider-aws/issues/7600)
*[ #829](https://github.com/hashicorp/terraform-provider-aws/issues/829)

**Comment by mbbush** [comment]
This is likely caused by the iam role being deleted before the application has finished deleting. Then the application doesn't have permission to clean up after itself. Try setting an explicit depends-on relationship.

**Comment by florianakos** [comment]
Just as @mbbush says. I managed to reproduce the issue with latest TF and AWS TF providers today, and was able to confirm via CloudTrail that the KDA APP tries to issue a `DeleteNetworkInterface` call that fails with error:

> You are not authorized to perform this operation. User: arn:aws:sts::XXXXXXXXXX:assumed-role/kda-role-test/1715670173340576933 is not authorized to perform: ec2:DeleteNetworkInterface on resource: arn:aws:ec2:eu-west-1: XXXXXXXXXX:network-interface/eni-03d70cf142fe6430d because no identity-based policy allows the ec2:DeleteNetworkInterface action.

This is ultimately cause by the fact that the IAM role the KDA app uses contains an attached policy (which includes permissions to manage ENI by itels) that gets destroyed before the KDA app can finish its own cleanup. This means an ENI resource will be left dangling which makes the Security Group destruction fail with error:

> Error: deleting Security Group (sg-abcdefghjklmnop): DependencyViolation: resource sg-abcdefghjklmnop has a dependent object

A possible fix is to add depends_on relation to the IAM role policy that's attached to the IAM role the KDA app uses:

```terraform
resource "aws_iam_role" "kda_role" { name = "kda-role-test" }

data "aws_iam_policy_document" "iam_permissions_with_ec2_eni_management_actions" { ... }

resource "aws_iam_role_policy" "attach_iam_permissions_with_ec2_eni_management_actions" {
  name   = "additional-role-policy"
  role   = aws_iam_role.kda_role.name
  policy = data.aws_iam_policy_document.iam_permissions_with_ec2_eni_management_actions.json
}

resource "aws_kinesisanalyticsv2_application" "kda" {
  name = "kda-test-application"
  service_execution_role = aws_iam_role.kda_role.arn
  depends_on = [ aws_iam_role_policy.attach_iam_permissions_with_ec2_eni_management_actions ]
  application_configuration { ... }
  ...
}
```

Another possible fix may be to create the IAM role and the associated IAM policy independently from the KDA app, so destruction of the KDA app does not happen in parallel with the destruction of the IAM resources it depends on for its cleanup.

**Comment by mbbush** [comment]
There are other resources that behave this way too. `aws_iot_topic_rule_destination` is one that comes to mind, but basically anything that creates and passes role that's given permission to manage ENIs.

**Comment by mbbush** [comment]
What happens if you specific the policy as an inline policy inside the
`aws_iam_role` resource? Maybe terraform is smart enough to see that the
kinesis app depends on the role, but not the policy.

**Comment by florianakos** [comment]
Yes, using an inline policy on the IAM role would be a sufficient workaround as well, to prevent errors when destroying a KDA app created by terraform.
